File: Darts/src/burlton/dartzee/code/screen/ScreenCache.java
package burlton.dartzee.code.screen;

import burlton.core.code.obj.SuperHashMap;
import burlton.core.code.util.Debug;
import burlton.dartzee.code.screen.ai.AIConfigurationDialog;
import burlton.dartzee.code.screen.game.DartsGameScreen;
import burlton.dartzee.code.screen.preference.PreferencesDialog;
import burlton.dartzee.code.screen.reporting.ConfigureReportColumnsDialog;
import burlton.desktopcore.code.bean.CheatBar;
import burlton.desktopcore.code.screen.BugReportDialog;
import burlton.desktopcore.code.screen.DebugConsole;
import burlton.desktopcore.code.util.DialogUtil;

import javax.swing.*;
import java.util.ArrayList;

public final class ScreenCache 
{
	private static SuperHashMap<String, DartsGameScreen> hmGameIdToGameScreen = new SuperHashMap<>();
	
	//Embedded screens
	private static SuperHashMap<Class<? extends EmbeddedScreen>, EmbeddedScreen> hmClassToScreen = new SuperHashMap<>();
	private static DartsApp mainScreen = null;
	
	//Dialogs
	private static HumanCreationDialog humanCreationDialog = null;
	private static AIConfigurationDialog aiConfigurationDialog = null;
	private static AboutDialogDarts aboutDialog = null;
	private static BugReportDialog bugReportDialog = null;
	private static PreferencesDialog preferencesDialog = null;
	private static PlayerImageDialog playerImageDialog = null;
	private static ConfigureReportColumnsDialog configureReportColumnsDialog = null;
	
	//Other
	private static DebugConsole debugConsole = null;
	
	public static <K extends EmbeddedScreen> K getScreen(Class<K> screenClass)
	{
		K scrn = (K)hmClassToScreen.get(screenClass);
				
		try
		{
			if (scrn == null)
			{
				scrn = screenClass.newInstance();
				hmClassToScreen.put(screenClass, scrn);
			}
		}
		catch (IllegalAccessException | InstantiationException iae)
		{
			Debug.stackTrace(iae);
			DialogUtil.showError("Error loading screen.");
		}
		
		return scrn;
	}
	
	public static DartsApp getMainScreen()
	{
		if (mainScreen == null)
		{
			mainScreen = new DartsApp(new CheatBar());
		}
		
		return mainScreen;
	}

	public static <K extends EmbeddedScreen> void switchScreen(Class<K> screenClass)
	{
		K screen = getScreen(screenClass);
		switchScreen(screen);
	}
	public static void switchScreen(EmbeddedScreen scrn)
	{
		switchScreen(scrn, true);
	}
	public static void switchScreen(EmbeddedScreen scrn, boolean reInit)
	{
		getMainScreen().switchScreen(scrn, reInit);
	}
	
	public static AIConfigurationDialog getAIConfigurationDialog()
	{
		if (aiConfigurationDialog == null)
		{
			aiConfigurationDialog = new AIConfigurationDialog();
		}
		
		aiConfigurationDialog.setLocationRelativeTo(getMainScreen());
		return aiConfigurationDialog;
	}
	
	public static HumanCreationDialog getHumanCreationDialog()
	{
		if (humanCreationDialog == null)
		{
			humanCreationDialog = new HumanCreationDialog();
		}
		
		humanCreationDialog.setLocationRelativeTo(getMainScreen());
		return humanCreationDialog;
	}
	
	public static PlayerManagementScreen getPlayerManagementScreen()
	{
		return getScreen(PlayerManagementScreen.class);
	}
	
	public static DebugConsole getDebugConsole()
	{
		if (debugConsole == null)
		{
			debugConsole = new DebugConsole();
		}
		
		return debugConsole;
	}
	
	public static AboutDialogDarts getAboutDialog() 
	{
		if (aboutDialog == null)
		{
			aboutDialog = new AboutDialogDarts();
		}
		return aboutDialog;
	}
	
	public static BugReportDialog getBugReportDialog() 
	{
		if (bugReportDialog == null)
		{
			bugReportDialog = new BugReportDialog();
		}
		return bugReportDialog;
	}
	
	public static PreferencesDialog getPreferencesDialog()
	{
		if (preferencesDialog == null)
		{
			preferencesDialog = new PreferencesDialog();
		}
		
		preferencesDialog.init();
		return preferencesDialog;
	}
	
	public static ArrayList<DartsGameScreen> getDartsGameScreens()
	{
		return hmGameIdToGameScreen.getValuesAsVector(true);
	}
	public static DartsGameScreen getDartsGameScreen(String gameId)
	{
		return hmGameIdToGameScreen.get(gameId);
	}
	public static void addDartsGameScreen(String gameId, DartsGameScreen scrn)
	{
		if (gameId.isEmpty())
		{
			Debug.stackTrace("Trying to cache GameScreen with no gameId.");
			return;
		}
		
		hmGameIdToGameScreen.put(gameId, scrn);
	}
	public static void removeDartsGameScreen(DartsGameScreen scrn)
	{
		hmGameIdToGameScreen.removeAllWithValue(scrn);
	}
	
	public static PlayerImageDialog getPlayerImageDialog()
	{
		if (playerImageDialog == null)
		{
			playerImageDialog = new PlayerImageDialog();
		}
		
		return playerImageDialog;
	}
	
	public static ConfigureReportColumnsDialog getConfigureReportColumnsDialog()
	{
		if (configureReportColumnsDialog == null)
		{
			configureReportColumnsDialog = new ConfigureReportColumnsDialog();
		}
		
		return configureReportColumnsDialog;
	}
	
	public static void exitApplication()
	{
		ArrayList<DartsGameScreen> openGames = getDartsGameScreens();
		int size = openGames.size();
		if (size > 0)
		{
			int ans = DialogUtil.showQuestion("Are you sure you want to exit? There are " + size + " game window(s) still open.", false);
			if (ans == JOptionPane.NO_OPTION)
			{
				return;
			}
		}
		
		System.exit(0);
	}
	
	public static void emptyCache()
	{
		hmClassToScreen.clear();
		
		humanCreationDialog = null;
		aiConfigurationDialog = null;
		aboutDialog = null;
		bugReportDialog = null;
		preferencesDialog = null;
		playerImageDialog = null;
		configureReportColumnsDialog = null;
	}
}

File: Darts/src/burlton/dartzee/code/screen/ScreenCache.kt
package burlton.dartzee.code.screen

import burlton.desktopcore.code.util.Debug
import burlton.dartzee.code.screen.ai.AIConfigurationDialog
import burlton.dartzee.code.screen.game.AbstractDartsGameScreen
import burlton.dartzee.code.screen.preference.PreferencesDialog
import burlton.dartzee.code.screen.reporting.ConfigureReportColumnsDialog
import burlton.dartzee.code.core.bean.CheatBar
import burlton.desktopcore.code.screen.DebugConsole
import burlton.desktopcore.code.util.DialogUtil
import javax.swing.JOptionPane

object ScreenCache
{
    private val hmGameIdToGameScreen = mutableMapOf<String, AbstractDartsGameScreen>()

    //Embedded screens
    private val hmClassToScreen = mutableMapOf<Class<out EmbeddedScreen>, EmbeddedScreen>()
    val mainScreen = DartsApp(CheatBar())

    //Dialogs
    private var humanCreationDialog: HumanCreationDialog? = null
    private var aiConfigurationDialog: AIConfigurationDialog? = null
    private var preferencesDialog: PreferencesDialog? = null
    private var configureReportColumnsDialog: ConfigureReportColumnsDialog? = null

    //Other
    val debugConsole = DebugConsole()

    fun getPlayerManagementScreen() = getScreen(PlayerManagementScreen::class.java)

    fun getDartsGameScreens() = hmGameIdToGameScreen.values

    fun <K : EmbeddedScreen> getScreen(screenClass: Class<K>): K
    {
        return hmClassToScreen.getOrPut(screenClass) { screenClass.getConstructor().newInstance() } as K
    }

    fun currentScreen() = mainScreen.currentScreen

    fun <K : EmbeddedScreen> switchScreen(screenClass: Class<K>)
    {
        val screen = getScreen(screenClass)
        switchScreen(screen)
    }

    fun switchScreen(scrn: EmbeddedScreen?, reInit: Boolean = true)
    {
        mainScreen.switchScreen(scrn, reInit)
    }

    fun getHumanCreationDialog(): HumanCreationDialog
    {
        if (humanCreationDialog == null)
        {
            humanCreationDialog = HumanCreationDialog()
        }

        humanCreationDialog!!.setLocationRelativeTo(mainScreen)
        return humanCreationDialog!!
    }

    fun getPreferencesDialog(): PreferencesDialog
    {
        if (preferencesDialog == null)
        {
            preferencesDialog = PreferencesDialog()
        }

        preferencesDialog!!.init()
        return preferencesDialog!!
    }

    fun getDartsGameScreen(gameId: String): AbstractDartsGameScreen?
    {
        return hmGameIdToGameScreen[gameId]
    }

    fun addDartsGameScreen(gameId: String, scrn: AbstractDartsGameScreen)
    {
        if (gameId.isEmpty())
        {
            Debug.stackTrace("Trying to cache GameScreen with no gameId.")
            return
        }

        hmGameIdToGameScreen[gameId] = scrn
    }

    fun removeDartsGameScreen(scrn: AbstractDartsGameScreen)
    {
        val keys = hmGameIdToGameScreen.filter { it.value == scrn }.map { it.key }
        keys.forEach { hmGameIdToGameScreen.remove(it) }
    }

    fun getConfigureReportColumnsDialog(): ConfigureReportColumnsDialog
    {
        if (configureReportColumnsDialog == null)
        {
            configureReportColumnsDialog = ConfigureReportColumnsDialog()
        }

        return configureReportColumnsDialog!!
    }

    fun exitApplication()
    {
        val openGames = getDartsGameScreens()
        val size = openGames.size
        if (size > 0)
        {
            val ans = DialogUtil.showQuestion("Are you sure you want to exit? There are $size game window(s) still open.", false)
            if (ans == JOptionPane.NO_OPTION)
            {
                return
            }
        }

        System.exit(0)
    }

    fun emptyCache()
    {
        hmClassToScreen.clear()

        humanCreationDialog = null
        aiConfigurationDialog = null
        preferencesDialog = null
        configureReportColumnsDialog = null
    }
}
